Added SignUp and Login

dmtrsn opened the pull request.

dmtrsn committed:
Added SignUp and Login

Max425 commented:
вот библиотека, для генирации документации к апи https://github.com/swaggo/swag

достаточно удобно, просто к хендлерам нужно аннотиции писать и сгенирируется свагер

Max425 commented:
Чистая архитектура - это ко 2 рк нужно делать. Сейчас вы рискуете, что вам снимут баллы, так как вы вперед бежите (такое было в прошлом году со многими командами, т.к. все сразу в чистой архитектуре стали делать) 

но может ничего и не сделают, зависит от препода, который принимать будет)

Max425 commented:
Хотя, если говорить про чистую архитектуру, то хеширование пароля все таки на уровне usecase стоит делать

также в модель юзера добавьте поле - соль

то есть не просто хешируйте пароль, но и для каждого пользователя будет какая-то рандомная строчка, которая будет использоваться при хешировании

вот пример, при создании пользователя у него создается соль, которая потом хешируется вместе с паролем
поэтому мы храним не просто хеш пароля, но ещё и засоленный
![image](https://github.com/user-attachments/assets/9300713e-d0bb-40f9-8dce-6fadb0471b93)

Max425 commented:
Структуру можете оставить как есть, если будут предъявлять, то скажте, что вы просто по папкам раскидали, у вас ещё не полноценно чистая архитектура (в чистой архитектуре на каждом слое используются разные модели) Но это вам уже только ко 2 рк надо будет делать

Max425 commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,92 @@
+
+// RegisterUser - обработчик регистрации пользователя
+func (h *UserHandler) RegisterUser(w http.ResponseWriter, r *http.Request) {
+	// Вопрос: как лучше поступать, елси пользователь авторизован через куки?
```
если у пользователя есть свежая куки, то просто 200 возвращайте

по хорошему вообще таких запросов не должно быть, так как на фронтенде для авторизованных пользователей должен работать редирект на страницу сайте

Max425 commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,92 @@
+
+	var user models.User
+	err := json.NewDecoder(r.Body).Decode(&user)
+	//TODO: реализовать валидацию
```
в моделе User можете сделать метод IsValid() bool
который будет проверять нужные вам поля
- поля не пустые
- имаил валидный (для этого можете библиотеку какую-нибудь использовать)

если невалидный, то отправляйте это на фронтенд, что какое-то поле некоректное

Max425 replied:
либо допишите в юзекейсах (пока что вы проверяете только, что не пустные)

Max425 commented on internal/response/response.go:
```diff
@@ -0,0 +1,21 @@
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(headerStatus)
+	json.NewEncoder(w).Encode(response)
+	// Вопрос: не лучше ли использовать http.Error()???
```
не критично, сейчас тоже хорошо сделали

Max425 commented on internal/usecase/userUsecase.go:
```diff
@@ -0,0 +1,33 @@
+}
+
+// RegisterUser - регистрация пользователя
+func (uc *UserUsecase) RegisterUser(user *models.User) error {
+	if user.Name == "" || user.Email == "" || user.Password == "" {
+		return errors.New("missing required fields")
+	}
+	return uc.repo.Save(user)
+}
+
+// AuthenticateUser - авторизация пользователя
+func (uc *UserUsecase) AuthenticateUser(user *models.User) (int, error) {
+	if user.Email == "" || user.Password == "" {
+		return 0, errors.New("missing required fields")
+	}
+	return uc.repo.Authenticate(user.Email, user.Password)
```
пароль вообще минимум 5 символов по хорошему
имаил тоже из одного символа не может быть

+ через либу проверяйте имаил на валидность (о чем писал выше)

Max425 replied:
проверять тут или в хендлерах - решать вам
я обычно такое в хендлерах проверяю, но и так тоже можете

Ivan-Skorokhodov committed:
moved hashing password to usecase and added salt

Ivan-Skorokhodov committed:
added funcs for valid login and registration fields

Ivan-Skorokhodov committed:
added handler for getting the list of courses

Ivan-Skorokhodov committed:
added logout

dmtrsn committed:
Added database interaction

Added database scripts for adding tables and testing data;
Added reg, auth, getting courses from db

dmtrsn committed:
Fixed ddl.sql

Ivan-Skorokhodov committed:
refactor and update ddl.sql and testing_data.sql

Ivan-Skorokhodov committed:
added sessions control with coockie in registration, login, logout using db

Ivan-Skorokhodov committed:
refactor and move params for connecting to db into .env

Max425 commented on cmd/server/main.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+// точка входа приложения
+func main() {
+	err := godotenv.Load("../../.env")
```
добавьте .gitignore файл и там .env напишите просто, а то вы его используете видимо, но в репозиторий его нельзя пушить

Max425 replied:
добавьте .gitignore файл и там .env напишите просто, а то вы его используете видимо, но в репозиторий его нельзя пушить

Max425 commented on cmd/server/main.go:
```diff
@@ -0,0 +1,51 @@
+func main() {
+	err := godotenv.Load("../../.env")
+	if err != nil {
+		log.Print(err)
+		log.Fatal("Ошибка загрузки .env файла")
```
log.Fatalf

Max425 commented on cmd/server/main.go:
```diff
@@ -0,0 +1,51 @@
+		log.Fatal("Ошибка загрузки .env файла")
+	}
+
+	host := os.Getenv("DB_HOST")
+	port := os.Getenv("DB_PORT")
+	user := os.Getenv("DB_USER")
+	password := os.Getenv("DB_PASSWORD")
+	dbname := os.Getenv("DB_NAME")
```
вынесите отдельно, и если какое-то поле пустое - ошибка, можно структуру под всё это дело создать

Max425 commented on internal/handlers/courseHandlers.go:
```diff
@@ -0,0 +1,37 @@
+// GetCourses - обработчик получения списка курсов
+func (h *CourseHandler) GetCourses(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		log.Print("method not allowed")
+		response.SendErrorResponse("method not allowed", http.StatusMethodNotAllowed, w)
```
log.Print - добавьте ещё что-нибудь, чтоб подробнее был, например, какая ручка была

Max425 commented on internal/models/course.go:
```diff
@@ -0,0 +1,12 @@
+package models
+
+type Course struct {
+	Id              int    `json:"-"`
```
поле id вообще оставьте, потом когда добавитье метод обновить, обычно это по id делается или удалять

Max425 commented on internal/repository/database.go:
сделайте репозиторий отдельный с методами для курса (в отдельный файл хотя бы вынесите)

Max425 commented on internal/handlers/courseHandlers.go:
```diff
@@ -0,0 +1,37 @@
+		return
+	}
+
+	bucketCourses, err := h.useCase.GetBucketCourses()
+	if err != nil {
+		log.Print(err)
+		response.SendErrorResponse(err.Error(), http.StatusNotFound, w)
+		return
+	}
```
ttp.StatusNotFound - если len(bucketCourses) == 0
а если ошибка - то 500 возвращайте

Max425 commented on internal/handlers/userHandlers.go:
```diff
@@ -0,0 +1,204 @@
+		return
+	}
+
+	err = h.useCase.RegisterUser(&user)
+	if err != nil {
+		log.Print(err)
+		response.SendErrorResponse(err.Error(), http.StatusBadRequest, w)
+		return
+	}
```
http.StatusBadRequest - ошибка пользователя (что-то не то передал)

а если не получилось зарегать или ешё чего, то это уже ошибка на вашей стороне - поэтому в подобных случаях нужно 500 возвращать

Max425 replied:
хотя, если не получилось зарегать по какой-то причине (email занят, например), то это можно 409 или ещё какой, как с фронтами решите
но слать всегда StatusBadRequest не совсем корректно

Max425 reviewed (commented):
Уже хорошо, правда тестов вообще нет, надо писать (табачные желательно сделайте)

Ivan-Skorokhodov committed:
made edits after the review

dmtrsn committed:
added cors and api/isAuthorized

dmtrsn committed:
added name field to register form

dmtrsn committed:
Fixed cookies and request to db from GetUserByCookie

Ivan-Skorokhodov committed:
added tests v1

Ivan-Skorokhodov committed:
little refactor and updated tests

Ivan-Skorokhodov committed:
updated coverage to 50%

Ivan-Skorokhodov committed:
modified readme

Ivan-Skorokhodov committed:
fixed bug, added comments and script for testing coverage
